### ISSUE TITLE
Program Trap

### ISSUE THREAD

**Issue by HansHoene** [HansHoene/Operating-Systems#4]
Inside MMU function, if an allocation request fails, the error counter is updated and the request is enqueued again into a synchronisation-proected queue.  If there is only one MMU handler, and it tries to enqueue into a full queue, the MMU handler thread will wait because of the semaphore until the queue is no longer full.  But the MMU handler thread is the only thread removing from the queue.  Therefore, since no requests in the queue are being serviced, the queue will forever be full, and the enqueue statement will stall forever.  The program stalls and does not do anything.

**Comment by HansHoene** [comment]
To fix, do not re-enqueue request.  Get rid of error counter and just return error from MMU.

**Comment by HansHoene** [comment]
Other idea: return error from MMU and update counter, but memory_malloc function catches error and continues to re-enqueue request.  This way, requesting thread is the thread that might stall while re-enqueing, but everything will be fine because MMU will continue servicing requests.

Pitfall = malloc must return error if requesting too much memory.

**Comment by HansHoene** [comment]
The issue was fixed with "other idea" option.  Unfortunately, the program stopped working and stalled.  Nothing would print.  I worked my way into the issue and discovered that threads were not inserting memory.  I used printf statements throughout the program to narrow down the error until I found that the line below was the culprit.  This line ensures that memory requests are not too large.  The problem is that memory->start is always less than memory->end.  To get the size of memory, one must actually do "memory->end - memory->start + 1".

`
	if (size > memory->start - memory->end + 1) {
		return -1;	// not enough space, bro
	}
`
